### PR TITLE
Add src/fs/cluster.hh

### PR DESCRIPTION
**Pull request by varqox**

### CODE REVIEW

File: src/fs/cluster.hh [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+struct cluster_range {
+	cluster_id_t beg;
+	cluster_id_t end; // exclusive
+};
+
+inline bool is_power_of_2(uint32_t x) noexcept {
+	return (x > 0 and (x & (x - 1)) == 0);
+}
+
+inline offset_t offset_to_cluster_id(offset_t offset, uint32_t cluster_size) noexcept {
```
Comment (tropuq) [inline src/fs/cluster.hh]:
```suggestion
inline cluster_id_t offset_to_cluster_id(offset_t offset, uint32_t cluster_size) noexcept {
```

**Review by tropuq — CHANGES_REQUESTED** [review]

File: src/fs/cluster.hh [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+using offset_t = uint64_t;
+using cluster_id_t = uint64_t;
+
+struct cluster_range {
+	cluster_id_t beg;
+	cluster_id_t end; // exclusive
+};
+
+inline bool is_power_of_2(uint32_t x) noexcept {
+	return (x > 0 and (x & (x - 1)) == 0);
```
Comment (rokinsky) [inline src/fs/cluster.hh]:
```suggestion
	return x > 0 and (x & x - 1) == 0;
```

File: src/fs/cluster.hh [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+}
+
+inline offset_t offset_to_cluster_id(offset_t offset, uint32_t cluster_size) noexcept {
+	assert(is_power_of_2(cluster_size));
+	return (offset >> __builtin_ctzll(cluster_size)); // should be 2 CPU cycles after inlining on modern x86_64
```
Comment (rokinsky) [inline src/fs/cluster.hh]:
```suggestion
	return offset >> __builtin_ctzll(cluster_size); // should be 2 CPU cycles after inlining on modern x86_64
```

File: src/fs/cluster_allocator.hh [hunk 4]
```diff
@@ -21,20 +21,21 @@
...
-    std::unordered_set<uint64_t> _allocated_blocks;
-    std::queue<uint64_t> _free_blocks;
+class cluster_allocator {
+    std::unordered_set<cluster_id_t> _allocated_clusters;
+    std::queue<cluster_id_t> _free_clusters;
 public:
-    explicit block_allocator(std::queue<uint64_t>);
-    uint64_t alloc();
-    void free(uint64_t addr);
+    explicit cluster_allocator(std::queue<cluster_id_t>);
```
Comment (varqox) [inline src/fs/cluster_allocator.hh]: Add some way to provide allocated clusters @wmitros

File: src/fs/cluster_allocator.hh [hunk 5]
```diff
@@ -21,20 +21,21 @@
...
-    std::unordered_set<uint64_t> _allocated_blocks;
-    std::queue<uint64_t> _free_blocks;
+class cluster_allocator {
+    std::unordered_set<cluster_id_t> _allocated_clusters;
+    std::queue<cluster_id_t> _free_clusters;
 public:
-    explicit block_allocator(std::queue<uint64_t>);
-    uint64_t alloc();
-    void free(uint64_t addr);
+    explicit cluster_allocator(std::unordered_set<cluster_id_t>, std::queue<cluster_id_t>);
```
Comment (varqox) [inline src/fs/cluster_allocator.hh]: Add names of parameters @wmitros

File: src/fs/cluster_allocator.hh [hunk 6]
```diff
@@ -21,20 +21,21 @@
 
 #pragma once
 
+#include "fs/cluster.hh"
```
Comment (varqox) [inline src/fs/cluster_allocator.hh]: Should we add newline between "..." and <...> includes?

File: src/fs/cluster.hh [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+
+inline bool operator!=(cluster_range a, cluster_range b) noexcept {
+	return not (a == b);
+}
+
+inline bool is_power_of_2(uint32_t x) noexcept {
+	return (x > 0 and (x & (x - 1)) == 0);
+}
+
+inline cluster_id_t offset_to_cluster_id(offset_t offset, uint32_t cluster_size) noexcept {
```
Comment (tropuq) [inline src/fs/cluster.hh]:
```suggestion
inline cluster_id_t offset_to_cluster_id(offset_t offset, unit_size_t cluster_size) noexcept {
```

File: src/fs/cluster.hh [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+	assert(is_power_of_2(cluster_size));
+	return (offset >> __builtin_ctzll(cluster_size)); // should be 2 CPU cycles after inlining on modern x86_64
+}
+
+inline offset_t cluster_id_to_offset(cluster_id_t cluster_id, uint32_t cluster_size) noexcept {
```
Comment (tropuq) [inline src/fs/cluster.hh]:
```suggestion
inline offset_t cluster_id_to_offset(cluster_id_t cluster_id, unit_size_t cluster_size) noexcept {
```

**Review by tropuq — CHANGES_REQUESTED** [review]

**Comment by varqox** [comment]
Renamed branch to fs-unit-headers -- see #47

**Pull request closed without merging**